pennylane_lightning/src/tests/Test_CompilerSupport.cpp

```diff
@@ -0,0 +1,35 @@
+#include <catch2/catch.hpp>
+
+#include <cassert>
+#include <exception>
+
+class TestExceptionInConstexprConstructor {
+  public:
+    explicit constexpr TestExceptionInConstexprConstructor(int t) {
+        if (t == 0) {
+            throw std::invalid_argument("Throw");
+        }
+    }
+};
+
+constexpr int testAssertInConstexprFunction(int t) {
+    assert(t != 0);
+    return t;
+}
+
+TEST_CASE("Test exception in a constexpr constructor") {
+    [[maybe_unused]] constexpr static auto t =
+        TestExceptionInConstexprConstructor(1);
+
+    // The following line must not be compiled
+    //[[maybe_unused]] constexpr static auto t1 =
+    // TestExceptionInConstexprConstructor(0);
```

Comment:
IS this to be uncommented?

Reply:
Yes, those lines must not be compiled (so it is failed to be compiled when uncommented). Still, it may be possible to add a SNIFAE-based solution to test whether an expression is compiled. Let me try it.

Reply:
I made several trials but failed to make it works. The problem is that this type of compilation failure is not appropriately reflected in the type system, e.g. when I use `decltype(testAssertInConstexprFunction(0))`, it just gives `int`. So I think we need to keep these lines.